android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableView.java

```diff
@@ -186,6 +234,59 @@ public void filterBy(String query)
 		this.isFiltered = true;
 	}
 
+	/**
+	 * Generate payload for `scroll` and `scrollend` events.
+	 *
+	 * @return KrollDict
+	 */
+	public KrollDict generateScrollPayload()
+	{
+		final KrollDict payload = new KrollDict();
+		final LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
+
+		// Obtain index for first visible row.
+		final View firstVisibleView =
+			layoutManager.findViewByPosition(layoutManager.findFirstVisibleItemPosition());
+		final TableViewHolder firstVisibleHolder =
+			(TableViewHolder) recyclerView.getChildViewHolder(firstVisibleView);
+		final TableViewRowProxy firstVisibleProxy = (TableViewRowProxy) firstVisibleHolder.getProxy();
+		final int firstVisibleIndex = firstVisibleProxy.getIndexInSection();
+		payload.put(TiC.PROPERTY_FIRST_VISIBLE_ITEM, firstVisibleIndex);
+
+		// Obtain scroll offset for content.
+		final KrollDict contentOffset = new KrollDict();
+		final TiDimension scrollOffsetXDimension = new TiDimension(scrollOffsetX, TiDimension.TYPE_WIDTH);
+		final TiDimension scrollOffsetYDimension = new TiDimension(scrollOffsetY, TiDimension.TYPE_HEIGHT);
+		contentOffset.put(TiC.EVENT_PROPERTY_X, scrollOffsetXDimension.getAsDefault(recyclerView));
+		contentOffset.put(TiC.EVENT_PROPERTY_Y, scrollOffsetYDimension.getAsDefault(recyclerView));
+		payload.put(TiC.PROPERTY_CONTENT_OFFSET, contentOffset);
+
+		// Approximate content size.
+		// NOTE: Due to recycling of views, we cannot calculate the true
+		// content size without loading all rows. The best we can do is an
+		// approximation based on first visible row.
+		final KrollDict contentSize = new KrollDict();
+		final TiDimension contentWidthDimension =
+			new TiDimension(firstVisibleView.getMeasuredWidth(), TiDimension.TYPE_WIDTH);
+		final TiDimension contentHeightDimension =
+			new TiDimension(firstVisibleView.getMeasuredHeight() * rows.size(), TiDimension.TYPE_HEIGHT);
+		contentSize.put(TiC.PROPERTY_WIDTH, contentWidthDimension.getAsDefault(recyclerView));
+		contentSize.put(TiC.PROPERTY_HEIGHT, contentHeightDimension.getAsDefault(recyclerView));
+		payload.put(TiC.PROPERTY_CONTENT_SIZE, contentSize);
+
+		// Obtain view size.
+		final KrollDict size = new KrollDict();
+		final TiDimension widthDimension =
+			new TiDimension(recyclerView.getMeasuredWidth(), TiDimension.TYPE_WIDTH);
+		final TiDimension heightDimension =
+			new TiDimension(recyclerView.getMeasuredHeight(), TiDimension.TYPE_HEIGHT);
+		size.put(TiC.PROPERTY_WIDTH, widthDimension.getAsDefault(recyclerView));
+		size.put(TiC.PROPERTY_HEIGHT, heightDimension.getAsDefault(recyclerView));
+		payload.put(TiC.PROPERTY_SIZE, size);
+
+		return payload;
+	}
+
```

Comment:
Yep, `visibleItemCount` isn't trivial with RecyclerView still trying to figure out an optimal way to get this

Reply:
If you think implementing `visibleItemCount` would cause performance issues (especially since scroll events will likely be spammed), then perhaps we should drop support for it. This might be okay since iOS doesn't support it either. If it's expensive, it might be better to provide a method to fetch it instead and for devs to call when a "scrollend" fires.